subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/DefaultComponentMetadataProcessor.java

```diff
@@ -179,7 +183,7 @@ public ModuleComponentResolveMetadata processMetadata(ModuleComponentResolveMeta
         if (!updatedMetadata.getStatusScheme().contains(updatedMetadata.getStatus())) {
             throw new ModuleVersionResolveException(updatedMetadata.getModuleVersionId(), () -> String.format("Unexpected status '%s' specified for %s. Expected one of: %s", updatedMetadata.getStatus(), updatedMetadata.getId().getDisplayName(), updatedMetadata.getStatusScheme()));
         }
-        return updatedMetadata;
+        return updatedMetadata.withDerivationStrategy(curStrategy);
```

Comment:
That's because of the non-cached rules: when they are processed, they play a "immutable to mutable" dance, which ends up losing the derivation strategy. Then the next time we do the comparison, we think it's using the same strategy when it's not.